spec/features/sign_up_spec.rb

```diff
@@ -0,0 +1,108 @@
+      end
+    end
+
+    context "mismatched password" do
+      scenario "doesn't add a user" do
+        expect{sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')}.to change{User.count}.by(0)
+      end
+
+      scenario "is empty" do
+        expect{sign_up('name', 'user', 'test@test.com', 'password', nil)}.to change{User.count}.by(0)
+      end
+
+      scenario "leaves user on /users page" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(current_path).to eq('/users/new')
+      end
+
+      scenario "throws an error message" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(page).to have_content("Password does not match the confirmation")
+      end
+
+      scenario "shows original email address in sign up form" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(find_field('email').value).to eq('test@test.com')
+      end
+
+      scenario "shows original name in sign up form" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(find_field('name').value).to eq('name')
+      end
+
+      scenario "shows original username in sign up form" do
+        sign_up('name', 'user', 'test@test.com', 'password', 'password_conf')
+        expect(find_field('username').value).to eq('user')
+      end
+    end
+  end
+
+  context "email address" do
+    scenario "is empty" do
+      expect{sign_up('name', 'user', nil, 'password', 'password')}.to change{User.count}.by(0)
+    end
+    scenario "not to have invalid format" do
+      expect{sign_up('name', 'user', 'testtest.com', 'password', 'password')}.to change{User.count}.by(0)
+    end
+  end
+
+  context "unique email" do
+    scenario "should be unique" do
+      User.create(name: 'name', username: 'user', email: 'test@test.com', password: 'password', password_conf: 'password')
+      expect{sign_up('name', 'user', 'test@test.com', 'password', 'password')}.to change{User.count}.by(0)
+    end
+    scenario "non unique email throws error" do
+      User.create(name: 'name', username: 'user', email: 'test@test.com', password: 'password', password_conf: 'password')
+      sign_up('name', 'user', 'test@test.com', 'password', 'password')
+      expect(page).to have_content("Email is already taken")
+    end
+  end
+
+  context "name" do
+    scenario "is empty" do
+      expect{sign_up(nil, 'user', 'test@test.com', 'password', 'password')}.to change{User.count}.by(0)
+    end
+    scenario "not to have invalid format" do
+      expect{sign_up('{)(£)}', 'user', 'testtest.com', 'password', 'password')}.to change{User.count}.by(0)
+    end
+  end
+
+  context "username" do
+    scenario "is empty" do
+      expect{sign_up('name', nil, 'test@test.com', 'password', 'password')}.to change{User.count}.by(0)
+    end
```

Comment:
I don't see how this tests that the username is empty. I don't think that's what you meant either - but this is why it's great to think about how you name scenarios carefully.